wandb/sdk/service/streams.py

```diff
@@ -0,0 +1,305 @@
+        # run_id = action.stream_id  # will want to fix if a streamid != runid
+        settings = action._data
+        thread = StreamThread(
+            target=wandb.wandb_sdk.internal.internal.wandb_internal,
+            kwargs=dict(
+                settings=settings,
+                record_q=stream._record_q,
+                result_q=stream._result_q,
+                port=self._port,
+                user_pid=self._pid,
+            ),
+        )
+        stream.start_thread(thread)
+        with self._streams_lock:
+            self._streams[action._stream_id] = stream
+
+    def _process_del(self, action: StreamAction) -> None:
+        with self._streams_lock:
+            stream = self._streams.pop(action._stream_id)
+            stream.join()
+        # TODO: we assume stream has already been shutdown.  should we verify?
+
+    def _process_drop(self, action: StreamAction) -> None:
+        with self._streams_lock:
+            stream = self._streams.pop(action._stream_id)
+            stream.drop()
+            stream.join()
+
+    def _finish_all(self, streams: Dict[str, StreamRecord], exit_code: int) -> None:
+        if not streams:
+            return
+
+        for sid, stream in streams.items():
+            wandb.termlog(f"Finishing run: {sid}...")  # type: ignore
+            stream.interface.publish_exit(exit_code)
+
+        streams_to_join = []
+        while streams:
+            for sid, stream in list(streams.items()):
+                poll_exit_resp = stream.interface.communicate_poll_exit()
+                if poll_exit_resp and poll_exit_resp.done:
+                    streams.pop(sid)
+                    streams_to_join.append(stream)
+                time.sleep(0.1)
+
+        # TODO: this would be nice to do in parallel
+        for stream in streams_to_join:
+            stream.join()
+
+        wandb.termlog("Done!")  # type: ignore
+
+    def _process_teardown(self, action: StreamAction) -> None:
+        exit_code: int = action._data
+        with self._streams_lock:
+            # TODO: mark streams to prevent new modifications?
+            streams_copy = self._streams.copy()
+        self._finish_all(streams_copy, exit_code)
+        with self._streams_lock:
+            self._streams = dict()
+        self._stopped.set()
+
+    def _process_action(self, action: StreamAction) -> None:
+        if action._action == "add":
+            self._process_add(action)
+            return
+        if action._action == "del":
+            self._process_del(action)
+            return
+        if action._action == "drop":
+            self._process_drop(action)
+            return
+        if action._action == "teardown":
+            self._process_teardown(action)
+            return
+        raise AssertionError(f"Unsupported action: {action._action}")
+
+    def _loop(self) -> None:
+        while not self._stopped.is_set():
+            # TODO: check for parent process going away
```

Comment:
Do you want to address this TODO in a different PR? should we track this?

Reply:
Added to:
https://wandb.atlassian.net/browse/WB-7684